[admin-menus] Display only components using Categories in List All Categories menu item

infograf768 opened the pull request:
Pull Request for Issue #13793

### Summary of Changes
Creates a new field type to display only components which have categories created.
### Testing Instructions
Create a new admin menu of type List All Categories
Before patch, we get

![screen shot 2017-01-29 at 09 29 18](https://cloud.githubusercontent.com/assets/869724/22402852/82b0d122-e605-11e6-8e79-69af721b2404.png)

After patch:

![screen shot 2017-01-30 at 11 47 39](https://cloud.githubusercontent.com/assets/869724/22420396/89a98a86-e6e2-11e6-858c-9e82e7a62d50.png)

### Documentation Changes Required
None

@Bakual @mbabker @izharaazmi @franz-wohlkoenig

infograf768 committed:
[admin-menus] Display only components using Categories

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on bd3d98c5677cc336684e976636a0aedd2d9b7c53<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13808">issues.joomla.org/tracker/joomla-cms/13808</a>.</sub>

laoneo commented on administrator/components/com_menus/models/fields/componentscategory.php:
```diff
@@ -0,0 +1,79 @@
+			->from('#__extensions as a')
+			->where('a.enabled >= 1')
+			->where('a.type =' . $db->quote('component'))
+			->join('INNER', '#__categories as b ON a.element=b.extension');
```
As b.extension can be "com_foo.bar" the check should probably be `a.element like b.extension +'%'`.

infograf768 replied:
@laoneo 
I have no component with which I could test this.
Please give me a link to an extension I can install with such behavior.

infograf768 replied:
Also, what would be the exact full code for this line as I am not familiar with it.

laoneo replied:
I just know that sections do exist for categories, so there must be extensions with it otherwise we wouldn't support them. YOu can simulate it by opening this url /administrator/index.php?option=com_categories&extension=com_content.foo

`->join('INNER', "#__categories as b ON b.extension like a.element +'%'");`

infograf768 replied:
That code will not work as the `+` can't be used.
What "could" work would be: `. $db->quoteName('a.element') . '%'`

But, as far as I know, the extension could be com_foo.bar in the categories table (to cope with categories sections and not in the extensions table.
In any case, I tested both

`->join('INNER', '#__categories as b ON b.extension LIKE "' . $db->quoteName('a.element') . '%"');`
and
`->join('INNER', '#__categories as b ON a.element LIKE "' . $db->quoteName('b.extension') . '%"');

I have no results, even for the extensions that have categories.

@Bakual
Need your input here

Bakual replied:
Lets go with the extension only for now. That menuitem will not work for extensions using com_foo.bar categories. Nothing is broken beside that particular custom menuitem for those specific extensions. It's a risk I would take and hopefully they pop up so it can be properly tested. 
If so, it would only need a change in the SQL which can be done with no B/C implications and could even be done in a patch release.

izharaazmi commented:
@infograf768 I remember few of my extensions that uses categories but
doesn't ship with a default category.

For this patch I'd rather group the list in two parts – one those you are
showing and another those you are hiding. See if you think it fit.

Bakual commented:
It's for sure the simplest approach allthough it will fail if there are no categories yet created.
However that is an easy fix for the user. He just needs to create a category first using the regular menu before he can create a custom category manager menuitem for that extension.

infograf768 committed:
docs missing

infograf768 committed:
Simplifying field type code. Using it also for Create Category. Modified
tip.

infograf768 commented:
I have simplified the new field type code. 
Added the same field type when using the admin menu item "Create a category"
Modified the tip for the field to explain the user that it is necessary to create at least one category for any component in order for it to be listed.

![screen shot 2017-01-31 at 07 04 10](https://cloud.githubusercontent.com/assets/869724/22454147/2405074e-e784-11e6-967d-7d0b6fe7d430.png)

@izharaazmi 
I think this should be enough and can go in.

Folks, please test again.

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 9c0e902f370aef938c89e82b3cf9310aefe5c78f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13808">issues.joomla.org/tracker/joomla-cms/13808</a>.</sub>

izharaazmi commented:
@infograf768

>  that it is necessary to create at least one category for any component in order for it to be listed.

If we are okay with this, then this is good to go.

infograf768 commented:
RTC. 

If someone uses multiple categories sections, we can modify to fit.



@rdeutz 

Can be merged<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13808">issues.joomla.org/tracker/joomla-cms/13808</a>.</sub>

izharaazmi commented:
@infograf768 Can you please list the core components that are horizontal in nature and their respective url format. I need to make ensure one of my coming proposal works.

What I can see so far:
```
com_categories  => option=com_categories&extension=com_foo.bar, 
com_fields      => option=com_fields&extension=com_foo.bar
```
Are there more?

Bakual commented:
com_fields uses `&context=com_foo.bar`, not `&extension=com_foo.bar`.

izharaazmi commented:
Oh well. Thanks. Are there more such features apart from categories and
fields?

Bakual commented:
I'm not aware of it.
